zcash_client_backend/src/data_api.rs

```diff
@@ -44,6 +47,129 @@ pub enum NullifierQuery {
+    pending_change_value: NonNegativeAmount,
+    value_pending_spendability: NonNegativeAmount,
+    unshielded: NonNegativeAmount,
+}
+
+impl AccountBalance {
+    pub fn new(
+        spendable_value: NonNegativeAmount,
+        pending_change_value: NonNegativeAmount,
+        value_pending_spendability: NonNegativeAmount,
+        unshielded: NonNegativeAmount,
+    ) -> Self {
+        Self {
+            spendable_value,
+            pending_change_value,
+            value_pending_spendability,
+            unshielded,
+        }
+    }
+
+    pub fn spendable_value(&self) -> NonNegativeAmount {
+        self.spendable_value
+    }
+
+    pub fn pending_change_value(&self) -> NonNegativeAmount {
+        self.pending_change_value
+    }
+
+    pub fn value_pending_spendability(&self) -> NonNegativeAmount {
+        self.value_pending_spendability
+    }
+
+    pub fn unshielded(&self) -> NonNegativeAmount {
+        self.unshielded
+    }
+
+    pub fn total(&self) -> NonNegativeAmount {
+        (self.spendable_value
+            + self.pending_change_value
+            + self.value_pending_spendability
+            + self.unshielded)
+            .expect("Account balance cannot overflow MAX_MONEY")
+    }
+}
+
+pub struct ScanProgress {
+    sapling_note_commitments_total: u64,
+    sapling_note_commitments_scanned: u64,
+}
+
+impl ScanProgress {
+    pub fn new(sapling_note_commitments_total: u64, sapling_note_commitments_scanned: u64) -> Self {
+        Self {
+            sapling_note_commitments_total,
+            sapling_note_commitments_scanned,
+        }
+    }
+
+    pub fn sapling_note_commitments_total(&self) -> u64 {
+        self.sapling_note_commitments_total
+    }
+
+    pub fn sapling_note_commitments_scanned(&self) -> u64 {
+        self.sapling_note_commitments_scanned
+    }
+}
+
+/// A type representing the potentially-spendable value of unspent outputs in the wallet.
+///
+/// The balances reported using this data structure may overestimate the total spendable value of
+/// the wallet, in the case that the spend of a previously received shielded note has not yet been
+/// detected by the process of scanning the chain. The balances reported using this data structure
+/// can only be certain to be unspent in the case that [`Self::chain_tip_height`] is equal to
+/// [`Self::fully_scanned_height]`, and even in this circumstance it is possible that a newly
+/// created transaction could conflict with a not-yet-mined transaction in the mempool.
+pub struct WalletSummary {
+    account_balances: BTreeMap<AccountId, AccountBalance>,
+    chain_tip_height: BlockHeight,
+    fully_scanned_height: BlockHeight,
+    progress: ScanProgress,
+    progress_since_fully_scanned: ScanProgress,
+}
+
+impl WalletSummary {
+    pub fn new(
+        account_balances: BTreeMap<AccountId, AccountBalance>,
+        chain_tip_height: BlockHeight,
+        fully_scanned_height: BlockHeight,
+        progress: ScanProgress,
```

Comment:
Overall scan progress is probably not a valuable metric, and will likely be removed.